### PR TITLE
Segmentation setup

### PR DESCRIPTION
**Pull request by GwydionJon**

### CODE REVIEW

File: adaptivefiltering/segmentation.py [hunk 1]
```diff
@@ -60,3 +65,199 @@ def __geo_interface__(self):
...
+        self.add_zoom_slider()
+        self.add_polygon_control()
+
+        # get a color picker to put on the right side of the map
+        self.color_pick = self.get_color_pick()
+        # setup the grid with a list of widgets
+        self.setup_grid([self.m, self.color_pick])
+
+    def get_bounds(self):
+        """takes the boundry coordinates of given dataset through the hexbin filter and returns them as a polygon
```
Comment (dokempf) [inline adaptivefiltering/segmentation.py:101]: `boundry`

File: adaptivefiltering/segmentation.py [hunk 2]
```diff
@@ -60,3 +65,199 @@ def __geo_interface__(self):
...
+
+        :return:
+            :param segmentation:
+                All current polygons in one segmentation object
+            :type segmentation: Segmentation
+
+
+        """
+        # polygon_list = [data["geometry"] for data in self.draw_control.data]
+        # polygon_list = [data for data in self.draw_control.data]
```
Comment (dokempf) [inline adaptivefiltering/segmentation.py:222]: This can be removed right?

File: adaptivefiltering/segmentation.py [hunk 3]
```diff
@@ -60,3 +65,199 @@ def __geo_interface__(self):
...
+            A segmentation object which is to be loaded.
+        :type segmentation: Segmentation
+
+
+        """
+        # imports the segmentation with a basic style to avoid issus after multiple savings and loadings.
+        # This should preserve the json structure indefinedtly as long as only our segmentations are loaded.
+        # draw_control_data_json = []
+
+        # draw_control_data_json = segmentation
```
Comment (dokempf) [inline adaptivefiltering/segmentation.py:241]: Remove?

File: adaptivefiltering/segmentation.py [hunk 4]
```diff
@@ -60,3 +65,199 @@ def __geo_interface__(self):
...
+        ]["boundary_json"]
+        # get the previous coordinate representation
+        boundary_coordinates = hexbin_geojson["coordinates"][0]
+        coordinates_mean = np.mean(boundary_coordinates, axis=0)
+        polygon_boundary = ipyleaflet.Polygon(
+            locations=boundary_coordinates, color="gray", opacity=0.9
+        )
+        return coordinates_mean, polygon_boundary
+
+    def add_zoom_slider(self):
```
Comment (dokempf) [inline adaptivefiltering/segmentation.py:132]: I am not sure it is worth to have these methods if we always want to call them exactly once. Comments?

File: adaptivefiltering/segmentation.py [hunk 5]
```diff
@@ -60,3 +65,199 @@ def __geo_interface__(self):
...
+            "main sidebar "
+
+            """,
+            ),
+        )
+
+    def draw_handler(target, action, geo_json):
+        """This can later be used to update the list of polygons on the side."""
+
+    def update_draw_control_color(self, change, *args, **kwargs):
```
Comment (dokempf) [inline adaptivefiltering/segmentation.py:199]: Are we expecting `args` and `kwargs`?

File: jupyter/demo.ipynb [hunk 6]
```diff
@@ -92,15 +149,23 @@
...
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
    "name": "python",
-   "version": ""
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.9.4"
```
Comment (dokempf) [inline jupyter/demo.ipynb:168]: I am bit confused why `nbstripout` did not remove all of this metadata

**Review by dokempf — CHANGES_REQUESTED** [review]
Thanks for the PR, I added some comments. It may look like many comments, but none of them are major issues. 

One thing I missed that was on the whiteboard is a `show()` method for the `Segmentation` class: What would be needed to add that? Would it have enough information to not need a data set for bounding box extraction?

**Comment by dokempf** [comment]
Additional comment: As Github Actions point out, the new dependency on `ipyleaflet` needs to be specified.

File: adaptivefiltering/segmentation.py [hunk 7]
```diff
@@ -60,3 +65,199 @@ def __geo_interface__(self):
...
+
+            "main sidebar "
+
+            """,
+            ),
+        )
+
+    def draw_handler(target, action, geo_json):
+        """This can later be used to update the list of polygons on the side."""
+
```
[the remark this answers is not in the document]
Comment (GwydionJon) [reply adaptivefiltering/segmentation.py:198]: Relic of me trying to sync the color with the polygons

File: jupyter/demo.ipynb (hunk 6 above)
Comment (GwydionJon) [reply jupyter/demo.ipynb:168]: I have no idea

File: adaptivefiltering/segmentation.py (hunk 4 above)
Comment (GwydionJon) [reply adaptivefiltering/segmentation.py:132]: I was once taught that the init of a class should be as compact as feasible, so i tried to wrap most things in a function to not clutter the init and also provide a way for a bit more documentation for each of these functions. 
One can now also easily minimize them to make the class a bit less crowed. 

that was my reasoning behind the functions.

File: jupyter/demo.ipynb [hunk 8]
```diff
@@ -79,7 +79,64 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "test_segmentation = test_map.return_polygon()\n",
+    "test_segmentation\n",
```
[the remark this answers is not in the document]
Comment (GwydionJon) [reply jupyter/demo.ipynb:111]: I will keep that in mind.

File: adaptivefiltering/pdal.py [hunk 9]
```diff
@@ -27,8 +27,9 @@ def execute_pdal_pipeline(dataset=None, config=None):
         The configuration of the PDAL pipeline, according to the PDAL documentation.
     :type config: dict
     :return:
-        A numpy array data structure containing the PDAL output.
-    :rtype: numpy.array
+        The full pdal pipeline object
+    :rtype: pipeline
```
Comment (dokempf) [inline adaptivefiltering/pdal.py:31]: When the exact type is not known, I think it is better to omit the `:rtype:` statement.

File: adaptivefiltering/pdal.py [hunk 10]
```diff
@@ -55,8 +56,11 @@ def execute_pdal_pipeline(dataset=None, config=None):
     # We are currently only handling situations with one output array
     assert len(pipeline.arrays) == 1
 
-    # Return the output array
-    return pipeline.arrays[0]
+    # Check whether we should return the full pipeline including e.g. metadata
```
Comment (dokempf) [inline adaptivefiltering/pdal.py:59]: Comment outdated

File: adaptivefiltering/segmentation.py [hunk 11]
```diff
@@ -54,9 +59,271 @@ def save(self, filename):
...
+        if dataset is None and segmentation["features"] == []:
+            raise Exception("an empty segmention was given.")
+
+        if dataset is None and segmentation is None:
+            # if no dataset or segmentation is given, the map will be centered at the SSC office
+            self.coordinates_mean = np.asarray([49.41745, 8.67529])
+            self.segmentation = None
+
+        if dataset is not None and type(dataset) is not DataSet:
```
Comment (dokempf) [inline adaptivefiltering/segmentation.py:117]: `type(x) is Foo` should generally be avoided in favor of `isinstance(x, Foo)` to allow subclassing by users. It is more pythonic.

File: adaptivefiltering/segmentation.py [hunk 12]
```diff
@@ -54,9 +59,271 @@ def save(self, filename):
...
+    def show(self):
+        """This will create a new InteractiveMap with bounds from the segmentation.
+             use
+                 map = self.show()
+                 map
+             to show and access the data of the new map or just
+                self.show() to show the map without interacting with it.
+        :param grid:
+             The grid object which holds the map and the right side interface
+         :type grid: ipyleaflet.grid
```
Comment (dokempf) [inline adaptivefiltering/segmentation.py:71]: Surprising indentation

File: adaptivefiltering/segmentation.py [hunk 13]
```diff
@@ -54,9 +59,271 @@ def save(self, filename):
...
+
+        # add polygon draw tool and zoom slider
+        self.add_zoom_slider()
+        self.add_polygon_control()
+
+        # setup the grid with a list of widgets
+        self.setup_grid([self.m])
+
+    def get_boundary(self, dataset):
+        """Takes the boundry coordinates of the  given dataset
```
Comment (dokempf) [inline adaptivefiltering/segmentation.py:158]: `boundry`

File: adaptivefiltering/segmentation.py [hunk 14]
```diff
@@ -54,9 +59,271 @@ def save(self, filename):
...
+            ],
+        )
+
+        # get the coordinates from the metadata:
+        # this gives us lat, lon but for geojson we need lon, lat
+        coordinates = json.loads(hexbin_pipeline.metadata)["metadata"][
+            "filters.hexbin"
+        ]["boundary_json"]["coordinates"]
+
+        # set up the segnemtation object to later load into the map
```
Comment (dokempf) [inline adaptivefiltering/segmentation.py:203]: `segnemtation`

File: adaptivefiltering/segmentation.py [hunk 15]
```diff
@@ -54,9 +59,271 @@ def save(self, filename):
...
+        """Takes the boundry coordinates of the  given dataset
+            through the pdal hexbin filter and returns them as a segmentation.
+
+        :param dataset:
+            The dataset from which the map should be displayed.
+            This needs to be in a valid georeferenced format. Eg.: EPSG:4326
+        :type dataset: Dataset
+
+        :return:
+        :param hexbin_segmentation:
```
Comment (dokempf) [inline adaptivefiltering/segmentation.py:167]: I think the `:param` here is wrong. It is describing `:return:`, right?

File: requirements-dev.txt [hunk 16]
```diff
@@ -7,3 +7,5 @@ pytest-cov
 sphinx
 sphinx_lfs_content
 sphinx_rtd_theme
+ipyleaflet
+ipywidgets
```
Comment (dokempf) [inline requirements-dev.txt:11]: `requirements-dev.txt` should only contain those dependencies that are *not* needed for the project itself, but for testing, documentation etc.

**Review by dokempf — CHANGES_REQUESTED** [review]
LGTM, almost all my comments are about documentation typos etc.

**Comment by dokempf** [comment]
Testing in CI revealed another issue that should be addressed.

**Pull request closed without merging**